### DISCUSSION TITLE
Optimizing cutting routes

### DISCUSSION THREAD

**Discussion by dgrogandesign** [arkypita/LaserGRBL#1619] [General]
Hi there. I really appreciate this software and have successfully used it to create lots of cool stuff. The only issue I am having is that the algorithm seems to be very inefficient in its route planning. it will travel from one point all the way to the end of the bed to reach another when it could have got there using a cut path instead of wasting time travelling. Is there any way to help the software to be more 'intelligent' in its route planning or is this just something I will have to put up with? Thanks.

**Comment by arkypita** [comment]
An optimization is already present.

When you open a raster image and select "vectorization" tool, you have a checkbox "optimize travel" and that is what is needed to find the fastest route between all path.

What you ask for is a bit more difficult to obtain, because for the cut you actually have to make sure that the order of the cut makes the holes first and then the outsides, otherwise the piece "falls" (detaches) before its internal parts have been cut.

It is possible, but not so immediate to implement.

**Comment by dgrogandesign** [comment]
Thansk for the reply. My question is specific to using SVG files, sorry i should have mentioned. I have included a screenshot of my window and you can see how many travel lines there are when it could be done without any at all apart from the ones needed to cut out the inner holes
<img width="976" alt="paths" src="https://user-images.githubusercontent.com/72930430/138590028-6ddaaf2d-37fc-4dae-a816-0610e856564a.png">
.

It seems like the order in which the cuts are made is not related to their proximity to each other which means that my cnc spends a lot of time darting from one end of the bed to the other which i find quite frustrating. Again this is certainly not a complaint, i am very grateful fort this software but it would be even better if it planned the routes better.

**Reply by kbrung** [reply]
I am interested in optimization os SVG cutting as well.

Regardless of cutting inside before outside being important, at the moment the cut is simply done in the order stated in the SVG, so leaving that optimization for later, is not putting us at any disadvantage. We also don't know if a given power level will result in a cut or simply an engraving, so with the SVG format we can't even decide if any parts will fall or not. We need knowledge of whether the path is a cut or not.

We have the opportunity to make a simple optimization that will provide a lot time savings. Possibly, also better precision as a result of not traveling all over the place.

There are 2 simple ways to fix this problem: SVG optimization or g-code optimization.

I suggest a postprocessor of g-code, that does an OK job as a start. As an SVG is turned into a sequence cut/engravings. Reordering it is a simple matter of finding a good enough minimization of total length traveled. I already made a simple and slow genetic algorithm to minimize traveling, so a better one shouldn't be too hard.

It can become even more advanced later on, by for instance reversing some of the cuts, to see if that will provide a better solution.

I can try to code a version that works on the generated g-code, if it will be of any value.

**Comment by Tribo-2dot0** [comment]
Em Inkscape você pode ditar a ordem de corte dividindo os caminhos do projeto em camadas e subcamadas... LaserGrbl sempre irá cortar primeiro as camadas inferiores.

**Reply by kbrung** [reply]
That is a good starting point, but not really feasible for my needs.

I cut out pieces of fabric that needs to be sewn together. I can't put every single stitching hole on a separate layer. My most simple design would have 411 layers.

A more advanced need is based on the fact that a cutout or stitching hole can be cut in either direction. As an example, I waste so much time going from (0,0) to (2,0), then cut back to (1,0) and the move to (4,0) to cut back to (3,0).

**Reply by Tribo-2dot0** [reply]
O gerenciador de objetos cria automaticamente uma sub-camada para cada objeto que obedece a mesma ordem das camadas (de baixo para cima)... se você fizer um círculo, copia-lo e for colando (fazendo cópias dele), cada cópia estará automaticamente na camada superior e dessa maneira você estará automaticamente ordenado a sequência de corte.

**Reply by kbrung** [reply]
Thank you for replying.

My designs often have a lot of symmetry where I make one half and then duplicate it, and then join the lines. This jumbles the order quite a bit. But the worst part is when I make stitching holes. I draw a line and apply dashes in the stroke selection dialog. Then I go to Extension and select Modify Path->Convert to dashes. This turns the stroke in to actual line segments. It is very unfortunate that it often makes the direction of the cut in a very counter productive way, meaning it over-travels, cuts backwards, then over-travels again.
